sbc_compassion/tools/import_letter_functions.py

```diff
@@ -158,46 +273,94 @@ def _find_qrcode(env, img, MULTIPAGE, file_,
+            test_img = cv2.imread(file_png)[int(top):bottom, int(left):right]
+            test_data = readEncode(test_img)
             # save it for zxing
             cv2.imwrite(file_png, img)
             # replace the image by the returned one
             data = _save_img(MULTIPAGE, file_, img)
         else:
-            cv2.imwrite(file_png, img[::-1, ::-1])
+            img = img[::-1, ::-1]
+            cv2.imwrite(file_png, img)
             qrcode = zx.decode(file_png, try_harder=True)
-    if qrcode is not None and 'XX' in qrcode.data:
-        partner_id, child_id = qrcode.data.split('XX')
-        child_id = env['compassion.child'].search(
-            [('code', '=', child_id)]).id
-        partner_id = env['res.partner'].search(
-            [('ref', '=', partner_id),
-             ('is_company', '=', False)]).id
 
-    else:
-        partner_id = None
-        child_id = None
+    partner_id, child_id = decodeBarcode(env, qrcode)
 
     dict_odoo = {
         'partner_id': partner_id,
         'child_id': child_id,
     }
     if test:
-        right = int(left)+int(width)
-        bottom = int(top) + int(height)
-        test_img = cv2.imread(file_png)[int(top):bottom, int(left):right]
-        with tempfile.NamedTemporaryFile(
-                suffix='.png') as temp_file:
-            cv2.imwrite(temp_file.name, test_img)
-            f = open(temp_file.name)
-            test_data = f.read()
-            f.close()
         dict_odoo.update({
-            'qr_preview': base64.b64encode(test_data)
+            'qr_preview': test_data
         })
     return dict_odoo, data, img
 
 
+def testDirectionQRcode(barcode):
+    """ Test if the direction of the QR code is correct
+    :param Barcode barcode: Data from zxing
+    :return: True if the qrcode is in the right direction
+    :rtype: bool
+    """
+    # vector between the two uppermost points (index 1 is
+    # the one at the upper leftmost corner)
+    top = np.array(barcode.points[2]) - np.array(barcode.points[1])
+    left = np.array(barcode.points[0]) - np.array(barcode.points[1])
+    diag = top + left
+    if diag[0] > 0 < diag[1]:
+        return True
+    else:
+        return False
+
+
+def decodeBarcode(env, barcode):
+    """ Split the barcode and return the id of the partner and the child.
+    If the partner is not found, return None (same for the child).
+    :param env: Odoo environment
+    :param Barcode barcode:
+    :return: partner.id, Child.id
+    :rtype: int, int
+    """
+    if barcode is not None and 'XX' in barcode.data:
+        barcode_split = barcode.data.split('XX')
+        if len(barcode_split) == 2:
+            partner_ref, child_code = barcode_split
+            child_id = env['compassion.child'].search(
+                [('code', '=', child_code)]).id
+            partner_id = env['res.partner'].search(
+                [('ref', '=', partner_ref),
+                 ('is_company', '=', False)]).id
```

Comment:
You can add a condition, `contract_ids != False` to be sure only one partner will be found (sometimes we have duplicates in production (due to employees creating twice the same partner).